delta/app/src/test/scala/ch/epfl/bluebrain/nexus/delta/routes/ResourcesRoutesSpec.scala

```diff
@@ -346,6 +349,51 @@ class ResourcesRoutesSpec extends BaseRouteSpec with IOFromMap with CatsIOValues
       }
     }
 
+    "fail to update a schema without resources/write permission" in {
+      aclCheck.subtract(AclAddress.Root, Anonymous -> Set(resources.write)).accepted
+      Put(s"/v1/resources/$projectRef/_/someId/updateSchema") ~> routes ~> check {
+        response.status shouldEqual StatusCodes.Forbidden
+        response.asJson shouldEqual jsonContentOf("errors/authorization-failed.json")
+      }
+    }
+
+    "fail to update a schema when providing a non-existing schema" in {
+      aclCheck.append(AclAddress.Root, Anonymous -> Set(resources.write)).accepted
+      thereIsAResourceWithSchema { id =>
+        Put(s"/v1/resources/$projectRef/wrongSchema/$id/updateSchema") ~> routes ~> check {
+          response.status shouldEqual StatusCodes.NotFound
+          response.asJson.hcursor.get[String]("@type").toOption should contain("InvalidSchemaRejection")
+        }
+      }
+    }
+
+    "fail to update schema when providing _ as schema" in {
+      thereIsAResourceWithSchema { id =>
+        Put(s"/v1/resources/$projectRef/_/$id/updateSchema") ~> routes ~> check {
+          response.status shouldEqual StatusCodes.BadRequest
+          response.asJson.hcursor.get[String]("@type").toOption should contain("IdenticalSchema")
+        }
+      }
+    }
+
+    "fail to update schema when providing current schema" in {
+      thereIsAResourceWithSchema { id =>
+        Put(s"/v1/resources/$projectRef/myschema/$id/updateSchema") ~> routes ~> check {
+          response.status shouldEqual StatusCodes.BadRequest
+          response.asJson.hcursor.get[String]("@type").toOption should contain("IdenticalSchema")
+        }
+      }
+    }
+
+    "succeed to update the schema" in {
+      thereIsAResourceWithSchema { id =>
+        Put(s"/v1/resources/$projectRef/otherSchema/$id/updateSchema") ~> routes ~> check {
+          response.status shouldEqual StatusCodes.OK
+          response.asJson.hcursor.get[String]("_constrainedBy").toOption should contain(schema3.id.toString)
+        }
+      }
+    }
```

Comment:
I think we need to assert that the side-effect happened

Reply:
I am not sure if this is the right way to think about it, but I thought here we are just checking that the response to the new endpoint is as expected. The whole thing including the fetch is done in integration?

Reply:
I think it's unclear what the `*RoutesSpec` should be testing in general, but whether unit or integration, it feels like the tests need to cover
1. parsing request
2. routing of request
3. serialisation of response

I would say that 2 is where the side effect test requirement comes from

Reply:
It's definitely fair to say this is likely covered in integration tests. I think our test purposes aren't clear

Reply:
If the routes were dummy routes, this test should still pass. But whether the second part (fetching the resource after the update) is working depends on the `resources` implementation that's passed to the routes in the test, and that is something that is tested elsewhere

Reply:
The issue is that you're not testing that the resources implementation was actually called. But given the state of our routes tests I think maybe we can just move on

Reply:
Maybe we can discuss about some guidelines to clarify these situations